### PR TITLE
CW complexes, cellular homology + a lot more

### PR DESCRIPTION
**Pull request by aljungstrom**
Here's a big one... This is a bunch of code primarily written by @loic-p and me about CW complexes and cellular homology. It's part of a bigger project with @mortberg and was previously in a private repo. I guess Anders should be reviewing this one. It contains:

1. Definitions of CW complexes, (finite) cellular maps and (finite) cellular homotopies. This includes things like (constructive versions of) the cellular approximation theorems. All of this can be found under `Cubical.CW`.
2. Chain complexes, chain homotopies and related constructions. We use finite counterparts of many of these constructions to keep things constructive.
3. Construction of the chain complex associated to a CW complex à la Buchholtz and Favonia. This includes a bunch of proofs concerning properties of the degree of (1) maps between spheres (2) maps between bouquets of spheres.
4. Perhaps **most importantly**: cellular homology and, in particular, its functoriality. Actually, this is concerns two homology theories: we first construct `Hˢᵏᵉˡ` which defines the homology of an explicitly given skeleton and then lift this to  `Hᶜʷ`, the homology of an arbitrary CW complex, i.e. a type which is merely equivalent to the colimit of some skeleton.  We have been careful to set things up so that it works for possibly infinite dimensional CW complexes and not just finite ones. You can find the main results in `Cubical.CW.Homology`

FYI, I also introduced a new definition of _<_ on the naturals and used a definition of Fin n in terms of it. I would suggest we use this as the primary definition of Fin n in the future as the original one caused several of the proofs in this PR to be ~3 times as long (transport hell). But let's leave that to a future discussion...

### CODE REVIEW

**Comment by aljungstrom** [comment]
@rwbarton @owen-fool maybe this is of interest to your formalisation project?

**Comment by rwbarton** [comment]
Nice! I saw your HoTT/UF abstract a couple weeks ago. I find it a bit surprising that it's possible to do cellular homology synthetically like this without having a homotopy-invariant homology theory first; I'll have to look into exactly how it works.

Of course, my burning question is: any prospects for proving the Hurewicz theorem for simply-connected finite CW complexes, in terms of this cellular homology?

**Comment by aljungstrom** [comment]
@rwbarton  We're hoping it should be easier but no promises yet. On the blackboard it felt like it should follow pretty easily given the appropriate theory about n-connected complexes. I'll get back to you if I make any progress.

File: Cubical/CW/Approximation.agda [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+  {f : realise C → realise D}
+  → finCellApprox D E g m → finCellApprox C D f m
+  → finCellApprox C E (g ∘ f) m
+fst (compFinCellApprox m {g = g} {f} (F , p) (G , q)) = composeFinCellMap m F G
+snd (compFinCellApprox m {C = C} {g = g} {f} (F , p) (G , q)) =
+  →FinSeqColimHomotopy _ _ λ x
+    → funExt⁻ p _
+     ∙ cong g (funExt⁻ q (fincl _ x))
+
+-- a finite cellular homotopies relative
```
Comment (mortberg) [inline Cubical/CW/Approximation.agda:75]: ?

File: Cubical/CW/Approximation.agda [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+                  (sym (funExt⁻ (cong fst
+                   ((elimFinβ {A = Goalₗ} FST-base p .snd) (injectSuc x))) c))
+
+-- first main result
+CWmap→finCellMap : (C : CWskel ℓ) (D : CWskel ℓ')
+  (f : realise C → realise D) (m : ℕ)
+  → ∥ finCellApprox C D f m ∥₁
+CWmap→finCellMap C D f m =
+  PT.map (λ {(g , hom)
+  → (record { fmap = fst ∘ g ; fcomm = λ r x → sym (hom r x) })
```
Comment (mortberg) [inline Cubical/CW/Approximation.agda:267]: copattern-match?

Comment (mortberg) [reply Cubical/CW/Approximation.agda:267]: Constructor?

File: Cubical/CW/Approximation.agda [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+
+-- converting the above to something on the right form
+private
+  pathToCellularHomotopy-main :
+    {C : CWskel ℓ} {D : CWskel ℓ'} (m : ℕ) (f-c g-c : finCellMap m C D)
+    (h∞' : FinSeqColim→Colim m ∘ finCellMap→FinSeqColim C D f-c
+         ≡ FinSeqColim→Colim m ∘ finCellMap→FinSeqColim C D g-c)
+    → ∥ finCellHomRel m f-c g-c (approx.h∞ m f-c g-c h∞') ∥₁
+  pathToCellularHomotopy-main {C = C} zero f-c g-c h∞' =
+    ∣ (record { fhom = λ {(zero , p) x → ⊥.rec (CW₀-empty C x)}
```
Comment (mortberg) [inline Cubical/CW/Approximation.agda:583]: record..

Comment (mortberg) [reply Cubical/CW/Approximation.agda:583]: And more later

File: Cubical/CW/Homotopy.agda [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+       ; (j = i0) → merid (inl tt) ((i ∧ l) ∨ (~ k))
+       ; (l = i1) → merid (((push (f .fmap (injectSuc (fsuc n)) x))
+                  ∙∙ (cong inr (H .fhom (injectSuc (fsuc n)) x))
+                  ∙∙ (sym (push (g .fmap (injectSuc (fsuc n)) x)))) j)
+                     (i ∨ (~ k))})
+     south
+
+-- Then, we connect the addition of MMmaps to the addition of abelian maps
+module bouquetAdd where
+  -- keeping imports here for now
```
Comment (mortberg) [inline Cubical/CW/Homotopy.agda:431]: Move

File: Cubical/Data/Fin/Inductive/Base.agda [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Function
+
+open import Cubical.Data.Nat
+open import Cubical.Data.Empty as ⊥
+open import Cubical.Data.Unit
+
+open import Cubical.Relation.Nullary
+
+-- inductive definition of <
```
Comment (mortberg) [inline Cubical/Data/Fin/Inductive/Base.agda:17]: Move this and related stuff to Nat

File: Cubical/Data/Fin/Inductive/Properties.agda [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                (sumFinGen _+A_ 0A (λ x → f (injectSuc x)))
+                (sumFinGen _+A_ 0A (λ x → g (injectSuc x)))
+                _+A_ assocA comm
+
+sumFinGenId : ∀ {ℓ} {A : Type ℓ} {n : ℕ} (_+_ : A → A → A) (0A : A)
+  (f g : Fin n → A) → f ≡ g → sumFinGen _+_ 0A f ≡ sumFinGen _+_ 0A g
+sumFinGenId _+_ 0A f g p i = sumFinGen _+_ 0A (p i)
+
+
+open import Cubical.Data.Fin.Base renaming (Fin to Fin*) hiding (¬Fin0)
```
Comment (mortberg) [inline Cubical/Data/Fin/Inductive/Properties.agda:141]: Separate file?

File: Cubical/HITs/SequentialColimit/Base.agda [hunk 7]
```diff
@@ -1,20 +1,36 @@
...
+  fpush : (n : Fin m) (x : X .obj (fst n))
+    → fincl (injectSuc n) x ≡ fincl (fsuc n) (X .map x)
+
+FinSeqColim→Colim : (m : ℕ) {X : Sequence ℓ} → FinSeqColim m X → SeqColim X
+FinSeqColim→Colim m (fincl n x) = incl x
+FinSeqColim→Colim m (fpush n x i) = push x i
+
+realiseSequenceMap : {C : Sequence ℓ} {D : Sequence ℓ'}
+  → SequenceMap C D → SeqColim C → SeqColim D
+realiseSequenceMap record { map = map ; comm = comm } (incl x) = incl (map _ x)
```
Comment (mortberg) [inline Cubical/HITs/SequentialColimit/Base.agda:34]: ??

Comment (mortberg) [reply Cubical/HITs/SequentialColimit/Base.agda:34]: Constructor?

File: Cubical/HITs/SequentialColimit/Properties.agda [hunk 8]
```diff
@@ -235,3 +242,272 @@ module _
...
+
+-- main result
+-- (todo: add more thy about colimits to get nicer proof)
+converges→ColimIso : ∀ {ℓ} {seq : Sequence ℓ} (n : ℕ)
+  → converges seq n
+  → Iso (obj seq n) (SeqColim seq)
+Iso.fun (converges→ColimIso {seq = seq} n e) = incl
+Iso.inv (converges→ColimIso {seq = seq} n e) = elimShifted seq n _ (shiftEqShifted seq n e)
+Iso.rightInv (converges→ColimIso {seq = seq} n e) = elimShifted seq n _
+  (record { inclP = λ {k} → paths k
```
Comment (mortberg) [inline Cubical/HITs/SequentialColimit/Properties.agda:340]: record...

File: Cubical/Algebra/CommRing/Localisation/InvertingElements.agda [hunk 9]
```diff
@@ -720,17 +720,11 @@ module DoubleLoc (R' : CommRing ℓ) (f g : fst R') where
...
-
-           ≡⟨ (λ i → f ^ l · (g ^ l · transportRefl r i · transportRefl (g ^ m) i)
-                           · (1r · transportRefl (f ^ m) i · transportRefl 1r i)) ⟩
-
-             f ^ l · (g ^ l · r · g ^ m) · (1r · f ^ m · 1r)
+      path : f ^ l · (g ^ l · r · (g ^ m))
+                   · (1r · (f ^ m) · 1r)
+           ≡ f ^ l · (g ^ l · r' · (g ^ n))
+                   · (1r · (f ^ n) · 1r)
+      path = f ^ l · (g ^ l · r · g ^ m) · (1r · f ^ m · 1r)
```
Comment (mortberg) [inline Cubical/Algebra/CommRing/Localisation/InvertingElements.agda:727]: @mzeuner check this out!

**Review by mortberg — CHANGES_REQUESTED** [review]
Cool stuff! Some minor comments

**Review by mortberg — APPROVED** [review]

**Pull request merged**